src/MaxText/layers/llama4.py

```diff
@@ -445,66 +512,22 @@ def __call__(
         previous_chunk=previous_chunk,
         bidirectional_mask=bidirectional_mask,
     )
-
-    attention_lnx = nn.with_logical_constraint(
-        attention_lnx, ("activation_batch", "activation_norm_length", "activation_embed")
-    )
+    attention_lnx = nn.with_logical_constraint(attention_lnx, self.activation_axis_names)
     intermediate_inputs = inputs + attention_lnx
 
     # Fully Connected
-    hidden_states = rms_norm(
-        num_features=intermediate_inputs.shape[-1],
-        dtype=cfg.dtype,
-        weight_dtype=cfg.weight_dtype,
-        name="post_self_attention_layer_norm",
-        kernel_axes=("norm",),
-        epsilon=cfg.normalization_layer_epsilon,
-    )(intermediate_inputs)
-    hidden_states = nn.with_logical_constraint(
-        hidden_states, ("activation_batch", "activation_norm_length", "activation_embed")
-    )
+    hidden_states = self.post_self_attention_layer_norm(intermediate_inputs)
+    hidden_states = nn.with_logical_constraint(hidden_states, self.activation_axis_names)
 
-    load_balance_loss = None
     if self.is_moe_layer:
-      # NOTE: the naming mismatch here is to ensure reverse compatibility with existing checkpoints.
-      # The `name` represents the weight name in JAX/checkpoints and so the class name
-      # is just for readability.
-      mlp_lnx = moe.get_routed_and_shared_moe(
-          name="Llama4MoEBlock_0",
-          config=cfg,
-          mesh=self.mesh,
-          kernel_init=initializers.nd_dense_init(1.0, "fan_in", "truncated_normal"),
-          kernel_axes=("embed", None),
-          dtype=cfg.dtype,
-          weight_dtype=cfg.weight_dtype,
-          quant=self.quant,
-      )(hidden_states)
+      mlp_lnx = self.moe_block(hidden_states)
     else:
-      mlp_lnx = mlp_block(
-          in_features=hidden_states.shape[-1],
-          intermediate_dim=cfg.mlp_dim,
-          activations=cfg.mlp_activations,
-          intermediate_dropout_rate=cfg.dropout_rate,
-          dtype=cfg.dtype,
-          weight_dtype=cfg.weight_dtype,
-          name="mlp",
-          config=cfg,
-          quant=self.quant,
-      )(hidden_states, deterministic=deterministic)
-    mlp_lnx = nn.with_logical_constraint(mlp_lnx, ("activation_batch", "activation_norm_length", "activation_embed"))
+      mlp_lnx = self.mlp(hidden_states, deterministic=deterministic)
+    mlp_lnx = nn.with_logical_constraint(mlp_lnx, self.activation_axis_names)
 
     layer_output = mlp_lnx + intermediate_inputs
-
-    layer_output = nn.Dropout(rate=cfg.dropout_rate, broadcast_dims=(-2,))(layer_output, deterministic=deterministic)
-
-    layer_output = nn.with_logical_constraint(
-        layer_output,
-        ("activation_batch", "activation_norm_length", "activation_embed"),
-    )
-
-    # NOTE: this is only needed for dropping MoE
-    if load_balance_loss is not None:
-      self.sow("intermediates", "moe_lb_loss", load_balance_loss)
```

Comment:
Why is this removed?

Reply:
Yeah, the load_balance_loss is always None, and also mentioned here: https://screenshot.googleplex.com/6dwFWhQwwkLmDot. Let me know if any questions :)